### PR TITLE
Implement SD card support for sunxi (allwinner d1)

### PR DESCRIPTION
**Pull request by jspngh**
Hi all,

I worked on a driver for the allwinner D1 SMHC peripheral a couple of months ago.
I saw the post of @orangecms in tosc-rs/mnemos#139 and (finally) decided to make the PR for oreboot.
I didn't know of the initial work in PR #636, sorry about the duplication.

The functionality in this PR is already in good shape, but probably needs some more polish and testing.
Currently I can already load some data from an SD card

<details>
<summary>Example output</summary>

```
oreboot 🦀
v 13
cpu_pll fa001000
cpu_axi 5000100
cpu_axi 5000100
peri0_ctrl was: f8216300
peri0_ctrl lock en
peri0_ctrl PLLs
peri0_ctrl set: f8216300
DDR3@792MHz
test OK
512M 🐏
Initialized card type: SDHC
Select response 700 - Stby
Bus width response 920 - Tran
Initialized card type: SDHC
Select response 700 - Stby
Bus width response 920 - Tran
Initialized card type: SDHC
Select response 700 - Stby
Bus width response 920 - Tran
{ɕ����� serial uart0 initialized
RISC-V vendor 5b7 arch 0 imp 0
timer init
reset init
ipi init
Set up extension CSRs
==== platform CSRs ====
   MXSTATUS  c06fe000
   MHCR      00000109
   MCOR      00000002
   MHINT     00004000
see C906 manual p581 ff
=======================
==== platform CSRs ====
   MXSTATUS  c06fe000
   MHCR      0000017f
   MCOR      00000003
   MHINT     0000610c
see C906 manual p581 ff
=======================
RustSBI version 0.3.2
.______       __    __      _______.___________.  _______..______   __
|   _  \     |  |  |  |    /       |           | /       ||   _  \ |  |
|  |_)  |    |  |  |  |   |   (----`---|  |----`|   (----`|  |_)  ||  |
|      /     |  |  |  |    \   \       |  |      \   \    |   _  < |  |
|  |\  \----.|  `--'  |.----)   |      |  |  .----)   |   |  |_)  ||  |
| _| `._____| \______/ |_______/       |__|  |_______/    |______/ |__|
Platform Name: T-HEAD Xuantie Platform
Implementation: oreboot version 0.1.0
[rustsbi] misa: RV64ACDFIMSUVX
[rustsbi] mideleg: ssoft stimer sext (0x000222)
[rustsbi] medeleg: ima ia lma la sma sa uecall ipage lpage spage (0x00b1f3)
[rustsbi] mie: msoft ssoft mtimer stimer mext sext (0x000aaa)
PMP0	 0x0 - 0x40000000 (A,R,W,X)
PMP1	 0x40000000 - 0x40200000 (A,R)
PMP2	 0x40200000 - 0x80000000 (A,R,W,X)
PMP3	 0x80000000 - 0x80200000 (A,R)
PMP4	 0x80200000 - 0xfffff800 (A,R,W,X)
PMP8	 0x0 - 0x0 (A,R,W,X)
DTB looks fine, yay!
Decompress 8105609 bytes from 0x44000004 to 0x40200000, reserved 25165824 bytes
Success, decompressed 20541736 bytes :)
Payload looks like Linux Image, yay!
DTB still fine, yay!
Enter supervisor at 40200000 with DTB from 41a00000
```
</details>

However, putting boot0 on the SD card doesn't detect it (BROM does not load it),
currently I'm still loading boot0 manually into SRAM (using JTAG).
I'm following [this page](https://linux-sunxi.org/Allwinner_Nezha#BSP_boot0_SPL) for the location of boot0 on the SD card,
so I *think* it's in the correct location and *some* eGON header is present there,
but still something is going wrong.

Also, output stops after `Enter supervisor at 40200000 with DTB from 41a00000`,
I think that's not supposed to happen? I was expecting to see output from Linux.

Feel free to give remarks, suggestions, try it out for yourself, etc.

### CODE REVIEW

**Comment by orangecms** [comment]
Hey, this is really cool! I'll go try this out at some point this week.
Admittedly, we have some improvements to do around the whole build setup.

Today, I merged the revival of a library that would declaratively describe the layout of an oreboot image.
Currently, we have a lot of thingd hardcoded, as you have seen. That'll be all reworked.
So far, the strategy is to get things to work first, then refactor etc.. :)

**Comment by jspngh** [comment]
> However, putting boot0 on the SD card doesn't detect it (BROM does not load it), currently I'm still loading boot0 manually into SRAM (using JTAG).

I found out why this wasn't working. The BL702 on the lichee rv dock pro implements the debug probe hardware
for JTAG access, but since these use the same pins (as for the connection to the SD card slot), this gave some issues.
The issues didn't manifest after *using JTAG* (to load bt0 into SRAM).

I replaced the BL702 firmware with one that only gives access to UART0.
It can now detect bt0 on the SD card and start from there, but there's still an issue.
Reading from the SD card in that scenario doesn't work for some reason (the card responds correctly until I read data).

**Comment by orangecms** [comment]
Right, that makes sense. What you can do is to detach the module from the carrier board, so you can be sure that the SD card pins are not fiddled with, and attach another USB serial to the UART header on the module itself. The design with JTAG sharing the SD pins can be really annoying for debugging. :sweat_smile: 
There are also micro SD breakout board that have another card slot, but also expose the pins, allowing for attaching an oscilloscope. If you have/are fimiliar with that kinda gear, that may also help.

**Comment by jspngh** [comment]
> Right, that makes sense. What you can do is to detach the module from the carrier board, so you can be sure that the SD card pins are not fiddled with, and attach another USB serial to the UART header on the module itself. The design with JTAG sharing the SD pins can be really annoying for debugging. sweat_smile There are also micro SD breakout board that have another card slot, but also expose the pins, allowing for attaching an oscilloscope. If you have/are fimiliar with that kinda gear, that may also help.

Yup okay, so it works when I use only the module :grin:.
But that's a pretty janky setup though, I'll try to find a way to use the carrier board.

File: src/mainboard/sunxi/nezha/bt0/src/main.rs [hunk 1]
```diff
@@ -571,6 +602,33 @@ extern "C" fn main() {
...
+        let skip = skip + (0x1 << 15); // 32K, the size of SPL/boot0
+        load(skip, ORE_ADDR, ORE_SIZE, &mut smhc).ok();
+
+        // Linux is behind oreboot + dtfs
+        let skip = skip + ORE_SIZE + DTF_SIZE;
+        load(skip, LIN_ADDR, LIN_SIZE, &mut smhc).ok();
+
+        // DTB is behind Linux
+        let skip = skip + LIN_SIZE;
+        load(skip, DTB_ADDR, DTB_SIZE, &mut smhc).ok();
```
Comment (rminnich) [inline src/mainboard/sunxi/nezha/bt0/src/main.rs:629]: what kind of diagnostic do we get when things fail in load()?

File: src/soc/src/sunxi/d1/smhc/card.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+impl CardStatus {
+    pub(crate) fn new(val: u32) -> Self {
+        Self(val)
+    }
+
+    /// Indicates if there was an error in the sequence of the authentication process.
+    pub fn auth_seq_error(&self) -> bool {
+        const BIT: u32 = 1 << 3;
+        (self.0 & BIT) == BIT
```
Comment (rminnich) [inline src/soc/src/sunxi/d1/smhc/card.rs:51]: A lot of times, in firmware, when things to wrong, you want to have reasonable info. Is there additional information in self.0 that would be useful to have available to the caller when things have gone wrong?

**Review by rminnich — DISMISSED** [review]
If this is working for you, now, I wonder if we shouldn't just get it in so others can try ti? This is still a research effort anyway.

File: src/mainboard/sunxi/nezha/bt0/src/main.rs (hunk 1 above)
Comment (jspngh) [reply src/mainboard/sunxi/nezha/bt0/src/main.rs:629]: I wonder what the right behavior is when loading fails.
Retrying, stopping or just continuing anyway.
At the very least I'll add a `println!` for the error, I'm doing that locally as well.

File: src/soc/src/sunxi/d1/smhc/card.rs (hunk 2 above)
Comment (jspngh) [reply src/soc/src/sunxi/d1/smhc/card.rs:51]: In the public interface for `CardStatus` I added functions for accessing the bits
that I thought could be interesting. It's not complete, but functions can be added as needed.
Or I can add a `pub fn raw(&self) -> u32` or just make `self.0` pub.

<details>
  <summary>List of card status bits</summary>
  
  ```
  Bit  Typ Clr Identifier         Meaning
  31   ERX  C  OUT_OF_RANGE       (1=Command's argument was out of range)
  30   ERX  C  ADDRESS_ERROR      (1=Misaligned address/block len mismatch)
  29   ERX  C  BLOCK_LEN_ERROR    (1=Wrong block length, bytelen mismatch)
  28   ER   C  ERASE_SEQ_ERROR    (1=Error in erase command sequence)
  27   ERX  C  ERASE_PARAM        (1=Wrong erease selection of write-blocks)
  26   ERX  C  WP_VIOLATION       (1=Write failed due to write-protection)
  25   SX   A  CARD_IS_LOCKED     (1=Card is locked by the host)
  24   ERX  C  LOCK_UNLOCK_FAILED (1=Lock/unlock sequence or password error)
  23   ER   B  COM_CRC_ERROR      (1=CRC check of previous command failed)
  22   ER   B  ILLEGAL_COMMAND    (1=Command not legal for the card state)
  21   ERX  C  CARD_ECC_FAILED    (1=Internal error correction failed)
  20   ERX  C  CC_ERROR           (1=Internal card controller error)
  19   ERX  C  ERROR              (1=General error, or Unknown error)
  18   -    -  Reserved (eMMC: UNDERRUN)
  17   -    -  Reserved (eMMC: OVERRUN) (eSD: DEFERRED_RESPONSE)
  16   ERX  C  CSD_OVERWRITE      (1=read-only CSD section doesn't match card
                                  content, or attempted to reverse the
                                  Copy/WP bits)
  15   ERX  C  WP_ERASE_SKIP      (1=partial erase error due to write-protect)
  14   SX   A  CARD_ECC_DISABLED  (1=Internal error correction wasn't used)
  13   SR   C  ERASE_RESET        (1=Erase sequence was aborted)
  12-9 SX   B  CURRENT_STATE      (00h..0Fh=state, see below)
  8    SX   A  READY_FOR_DATA     (1=Ready/buffer is empty)
  7    EX   C  SWITCH_ERROR       (1=SWITCH command refused, MMC only)
  6    -    -  Reserved/Unspecified (description is left blank)
  5    SR   C  APP_CMD            (1=Card will expect ACMD)
  4    -    -  Reserved for SD I/O Card
  3    ER   C  AKE_SEQ_ERROR      (1=Authentication Sequence Error)
  2    -    -  Reserved for application specific commands
  1-0  -    -  Reserved for manufacturer test mode
  ```
</details>

**Comment by jspngh** [comment]
> If this is working for you, now, I wonder if we shouldn't just get it in so others can try ti? This is still a research effort anyway.

There's still some improvements I would like to make. Primar

**Comment by jspngh** [comment]
Oops, accidentally closed the PR.

**Comment by orangecms** [comment]
> > If this is working for you, now, I wonder if we shouldn't just get it in so others can try ti? This is still a research effort anyway.
> 
> There's still some improvements I would like to make. For instance, the clock settings need to be taken into account, which should speed up the loading process (it's pretty slow at the moment).

I've marked the PR as a draft, accordingly. Just undraft it when you think it's good. :)

**Comment by jspngh** [comment]
Sorry for the delay, it should be ready to merge now.

**Review by orangecms — APPROVED** [review]
Nice, thank you so, so much! :pray:

**Pull request merged**